goodkey/good_key.go

```diff
@@ -351,3 +379,54 @@ func checkSmallPrimes(i *big.Int) bool {
 	result.GCD(nil, nil, i, smallPrimesProduct)
 	return result.Cmp(big.NewInt(1)) != 0
 }
+
+// Returns an error if the modulus n is able to be factored into primes p and q
+// via Fermat's factorization method. This method relies on the two primes being
+// very close together, which means that they were almost certainly not picked
+// independently from a uniform random distribution. Basically, if we can factor
+// the key this easily, so can anyone else.
+func checkPrimeFactorsTooClose(n *big.Int, rounds int) error {
+	// Pre-allocate some big numbers that we'll use a lot down below.
+	one := big.NewInt(1)
+	bb := big.NewInt(0)
+
+	// Any odd integer is equal to a difference of squares of integers:
+	//   n = a^2 - b^2 = (a + b)(a - b)
+	// Any RSA public key modulus is equal to a product of two primes:
+	//   n = pq
+	// Here we try to find values for a and b, since doing so also gives us the
+	// prime factors p = (a + b) and q = (a - b).
+
+	// We start with a close to the square root of the modulus n, to start with
+	// two candidate prime factors that are as close together as possible and
+	// work our way out from there. Specifically, we set a = ceil(sqrt(n)), the
+	// first integer greater than the square root of n. Unfortunately, big.Int's
+	// built-in square root function takes the floor, so we have to add one to get
+	// the ceil.
+	a := big.NewInt(0)
+	a.Sqrt(n).Add(a, one)
+
+	// We calculate b2 to see if it is a perfect square (b^2), and therefore b is an
+	// integer. Specifically, b2 = a^2 - n.
+	b2 := big.NewInt(0)
+	b2.Mul(a, a).Sub(b2, n)
+
+	for i := 0; i < rounds; i++ {
+		// To see if b2 is a perfect square, we take its square root, square that,
+		// and check to see if we got the same result back.
+		bb.Sqrt(b2).Mul(bb, bb)
+		if b2.Cmp(bb) == 0 {
+			// b2 is a perfect square, so we've found integer values of a and b,
+			// and could easily compute p and q as their sum and difference.
+			bb.Sqrt(bb)
+			p := big.NewInt(0).Add(a, bb)
+			q := big.NewInt(0).Sub(a, bb)
+			return fmt.Errorf("public modulus n = pq factored into p: %s; q: %s", p.String(), q.String())
```

Comment:
Nit: `Errorf` will call `String()` for you.

```suggestion
			return fmt.Errorf("public modulus n = pq factored into p: %s; q: %s", p, q)
```

Reply:
Huh! Weird that it works for big.Int but not for big.Float:
> String must be called explicitly, Float.Format does not support %s verb.

I'd seen that warning under big.Float.String() and just assumed it applied to both.

Reply:
That's odd. The `%s` verb is based on the `Stringer` interface, and you can't opt out of interface implementation if you happen to have a function with the right signature.

https://pkg.go.dev/fmt
> 5. If an operand implements method String() string, that method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any).

Perhaps that comment on Float.String is specifically with reference to Float.Format and isn't talking about Printf/Errof?

Reply:
Nope, it really truly doesn't work for `big.Float`: https://go.dev/play/p/a9OnZ31HC4i
But it does for `big.Int`. I am very confused.

Reply:
Anyway I've done this, since the formatting clearly does work for big.Int.

Reply:
Quick note on the behavior here:
https://pkg.go.dev/fmt
> 2. If an operand implements the Formatter interface, it will be invoked. In this case the interpretation of verbs and flags is controlled by that implementation.

The Formatter interface takes precedent over stringer if defined (Spoiler alert it is: https://cs.opensource.google/go/go/+/refs/tags/go1.17.8:src/math/big/ftoa.go;drc=770fac4586a9c8c4647cb6ff79443ec246ae32c4;l=465)

And the Formatter not containing a case to handle %s matches the builtin floatX classes.  Interestingly enough the builtin floatX lack the String() function